### ISSUE TITLE
Assign journalists to specific sources

### ISSUE THREAD

**Issue by garrettr** [freedomofpress/securedrop#1355]
It would be useful to be able to assign a specific source to a given journalist, so that journalist (and the other journalists) know who is handling them.

In the ORM, it's trivial to create a foreign key between `Source` and `Journalist`. It's also easy to display the assigned Journalist in the Sources and Source views on the Document Interface.

Since journalists may be assigned, then re-assigned, it might be useful to track and display the history of assignments in the Source view (similar to how GitHub displays assignees to issues, for example).

In the simplest implementation of this feature, any Document Interface user would be able to assign (or re-assign) a journalist to a source. It would be more complicated, but do-able, to base this capability on some kind of permissions system (e.g. only "admins" or "editors" can assign/re-assign sources or journalists). I am not sure if permissions for assignment would be useful, so I am inclined to implement the simplest version of this first, and implement/track permissions in a follow-up issue if that feature is requested.

It's an open question whether it would be useful to let Sources know who is assigned to them. My inclination is to say that this would not be useful, and could potentially be used to target or harm specific journalists, so I would prefer not to implement this unless somebody can make a strong case for doing so.

**Comment by fowlslegs** [comment]
> It's an open question whether it would be useful to let Sources know who is assigned to them. My inclination is to say that this would not be useful, and could potentially be used to target or harm specific journalists, so I would prefer not to implement this unless somebody can make a strong case for doing so.

This should certainly not be the default. Hesistant to even add it as an optional setting. A journalist can always choose to reveal their identity in a message to the source.

**Comment by heartsucker** [comment]
A source who frequently checks also might see themselves repeatedly
reassigned which could be off putting or might give the impression of a
lack of organization among the news org. Not to mention if they know this
feature exists and see none, this will know they are unassigned which could
be even worse.

I agree that sources should not know what journalist they are assigned to.
Am 11.07.2016 8:59 nachm. schrieb "Noah Vesely" <notifications@github.com>:

**Comment by fowlslegs** [comment]
This is not the exact perfect place for this thought, but when considering adding more and more {meta,}data that will be stored to the app server, I think we might want to consider how we could leverage a journalist client to deprivilege this information from the app server. A solution might include some multi-party computation scheme where state such as which journalist is assigned to which source can be synced across journalist clients through the app server, without the app server being able to gain knowledge. I'm not suggesting this is how we should first implement this feature--a straightforward implementation might be best at first--just that this may be a way to help preserve the security of our system, while continuing to provide more rich functionality in our application.

**Comment by redshiftzero** [comment]
This was merged in #1491

**Comment by heartsucker** [comment]
Reopening in relation to #2841.

**Comment by heartsucker** [comment]
Open questions about how assignment should work:

- Who is authorized to assign sources? Only admins?
- What happens to a source who is unassigned?
- How does this relate to a non-shared inbox? (i.e., on inbox per journalist)
  - What happens to past messages when someone is reassigned?
- Should assignment be 1:1 or 1:N? What does the inbox look like by default?
- What if an org doesn't want source assignment and only one shared inbox? What if an org doesn't want a shared inbox at all? Do we have any use research about this?

**Comment by eloquence** [comment]
Judging by the screenshots in #1359, I agree this implementation would already address significant pain points. I would propose some slight modifications:

- as noted above, support 1:N source:journalist relationships
- limit assignment actions to "(un-)assign yourself" to address most common concern established through user research (lack of tracking of which member of the team has looked at a given submission) while keeping UI impact minimal
- display date of the assignment(s), since it's useful for orgs (they often track this information in spreadsheets right now).

Before implementation, I would suggest building an interactive wireframe of the proposed UI, and working with the UX team on some light validation with users who are willing to participate. 

If we do implement such a simple workflow improvement, I do see it as decoupled from #2841. If per-daughter-organization inboxes were implemented, each would still potentially be available to multiple users, and each would retain the assignment feature in its organizational context, so I see no conflict in a staged deployment.

**Comment by huertanix** [comment]
@heartsucker Some answers to your Qs below:

_Who is authorized to assign sources? Only admins?_

An admin's role in my experience has been less editorial and more IT, so I don't think admins are the folks who would create those assignments. My hunch is that journalists would know each other's assignment desks enough to know how to assign themselves or each other. Without a built-in means of communicating that assignment (my brief glance at #1359 does not include any kind of notification system), there's a risk a journalist might manually let their colleague know of their assignment through a less secure medium (email or Facebook Messenger). That workflow can be addressed as part of the SecureDrop training and documentation, but leaves wiggle room for mistakes.

_What happens to a source who is unassigned?_

This is a good question; My understanding of this feature's need is more tagging/labeling than access control, so I'd imagine basically nothing beyond the label changing.

_How does this relate to a non-shared inbox? (i.e., on inbox per journalist)_

Echoing my previous thoughts, I figure this wouldn't need to separate inboxes to meet the need of organizing/labeling drops.

_What happens to past messages when someone is reassigned?_

I figure nothing beyond the changing of the label.

_Should assignment be 1:1 or 1:N? What does the inbox look like by default?_

It's definitely possible to see multiple people assigned to one source in larger newsrooms, I tend to lean towards 1:N.

_What if an org doesn't want source assignment and only one shared inbox? What if an org doesn't want a shared inbox at all? Do we have any use research about this?_

The feedback I've heard during training has generally been less of a concern about who can see what or getting scooped so much as having to remember which source a journalist was talking to. There's been more formal user research around this recently; some of the researchers may have good feedback on the needs surfaced in their discussions.

**Comment by eloquence** [comment]
> *there's a risk a journalist might manually let their colleague know of their assignment through a less secure medium*

That might be another reason to stick to self-assignment, at least initially.

**Comment by heartsucker** [comment]
I have a branch pushed with a draft of this. Can I get a UI/UX check before I write all the tests and polish this?

![screenshot_2018-09-24_22-15-08](https://user-images.githubusercontent.com/3998464/45976553-6258d500-c047-11e8-817f-6e06b9a26a21.png)

![screenshot_2018-09-24_22-14-57](https://user-images.githubusercontent.com/3998464/45976562-6684f280-c047-11e8-9a75-d218def14b55.png)

**Comment by eloquence** [comment]
@heartsucker @freedomofpress/securedrop-ux 

What happens if, in the first screenshot, I click on "dellsberg" or "unassigned" -- does it take me to the assignment screen? 

What happens in the per-source view if the source is not assigned yet?

A few observations, looking forward to input from UX team members:
- We have an existing problem in the UI where it's not always clear if something is an indicator or a button. We should be careful to not add to that confusion.
- There is a cognitive clash in the second screenshot between "dellsberg" and "(unassigned)" -- because the dropdown stands out, it would be easy to misinterpret the status as being unassigned.
- We should avoid adding a lot of verbiage if only a subset of users are expected to use this feature, so we may want to opt for a more compact view until/unless an assignment has been made.
- I would suggest quoting the codename ("progressive shindig") and perhaps the username in the output message, because dictionary words can lead to otherwise confusing results.

As we get more systematic about UX research, I do think we should attempt to get some evidence from users that this would help them to manage submissions before merging an iteration of this feature. Nonetheless, glad to see progress on this front!

**Comment by heartsucker** [comment]
> What happens if, in the first screenshot, I click on "dellsberg" or "unassigned" -- does it take me to the assignment screen?

Nothing happens. Since the docs/messages buttons don't do anything, neither does this.

> What happens in the per-source view if the source is not assigned yet?

Text says "this source is unassigned."

> There is a cognitive clash in the second screenshot between "dellsberg" and "(unassigned)" -- because the dropdown stands out, it would be easy to misinterpret the status as being unassigned.

Yeah, I was going to clean that up later. Getting it to correctly display the already-assigned journalist was annoying, so I left it for prototyping.

> We should avoid adding a lot of verbiage if only a subset of users are expected to use this feature, so we may want to opt for a more compact view until/unless an assignment has been made.

Fair. I can tighten it down and post a new screenshot.

> I would suggest quoting the codename ("progressive shindig") and perhaps the username in the output message, because dictionary words can lead to otherwise confusing results.

Right.

**Comment by heartsucker** [comment]
Minor update.

![screenshot_2018-09-25_11-43-31](https://user-images.githubusercontent.com/3998464/46006661-56f1c200-c0b8-11e8-9058-aedeb463f9a1.png)

**Comment by eloquence** [comment]
Much cleaner already, thanks; look forward to any follow-up screenshots. I still think this is pretty heavy on the source view and suspect it would be less cluttered if we can somehow come up with a nice progressive disclosure pattern that works without JS.  Will also ask Nina to weigh in.

**Comment by ninavizz** [comment]
Off the bat, it seems like Notifications would be a highly desirable (if not a dependent) parallel functionality... but otherwise, this sounds like a great idea!

**Comment by heartsucker** [comment]
@ninavizz here for the list of all source: https://github.com/freedomofpress/securedrop/issues/1355#issuecomment-424109821

I think we shouldn't do notifications for this since for a first pass. I think it's enough to have the basic overview on the homepage.

**Comment by ninavizz** [comment]
@heartsucker Agreed re: Notifications. That's how I'd split-up the feature for Beta on the Journalist Workstation, actually. Notifications are their own problem to nicely solve for. 

I've conferred with Erik, and tomorrow afternoon we'll run some quick wireframes I'll do today, by a user for feedback. Will post those here for team feedback, this afternoon PST. w00t! Thank you for reminding me in today's standup call!

**Comment by ninavizz** [comment]
Hasty prototype. Click anywhere on the screen to see hot-spots. Invision sucks at hover states, so you have to click to view the hover in the dropdown. There's also a little grid in the lower-right, and if you click on that it shows all screens. https://invis.io/NSOD4U8TCQB#/323206873_All_Sources

Thoughts, feedback?

**Comment by eloquence** [comment]
Thanks for the quick sketch!

If I understand correctly, you're intentionally omitting the information about assignments in the list view. Personally, I'm not sure that's a good idea -- while I worry about cluttering the list view as well, the fact that something has already been assigned is important information because it could enable some triage of incoming submissions (either via self-assignment or through a person on staff managing the assignments).

**Comment by ninavizz** [comment]
Thank you, Erik! No, that was not intentional—my bad! I'll go ahead and update that quickly...

**Comment by heartsucker** [comment]
Yeah, agree that it's kinda important to have on there if for no other reason than to be able to say "Steve is on vacation; I better respond to his sources."

**Comment by heartsucker** [comment]
But yeah, I can get at this tomorrow. Today is a holiday in Germany (and yet here I am on Github).

**Comment by ninavizz** [comment]
Did some testing with @eloquence today and had one user look at this feature. Have yet to sync with him on any thoughts following our testing, but to close the loop on this concept at least, I'll do the outstanding mockup on this tomorrow PM PST. Go enjoy the DE holiday! :)

**Comment by eloquence** [comment]
We'll consider this in the context of feature work on the SecureDrop Workstation, keeping open for now.
